File: test/2/MergeSort.php
<?php
class MergeSort{

    public function sort(&$arr) {
        $r = count($arr) - 1;
        $this->sortRecursion($arr, 0, $r);
    } 

    public function sortRecursion(&$arr, $l, $r) {

        // 递归终止条件
        if ($l >= $r) {
            return ;
        }

        $mid = $l + (int) (($r - $l) / 2);
        
        $this->sortRecursion($arr, $l, $mid);
        $this->sortRecursion($arr, $mid+1, $r);

        $this->merge($arr, $l, $mid, $r);
    }   

    public function merge(&$arr, $l, $mid, $r) {

        $i = $l;
        $j = $mid + 1;

        $tmp = [];
        while($i <= $mid && $j <= $r) {
            if ($arr[$i] < $arr[$j]) {
                $tmp[] = $arr[$i];
                $i++;
            } else {
                $tmp[] = $arr[$j];
                $j++;
            }
        }

        if ( $j <= $r ) {
            $i = $j;
            $mid = $r;
        }
        // 还有剩余不分
        while( $i<=$mid ) {
            $tmp[] = $arr[$i];
            $i++;
        }

        foreach($tmp as $key => $val) {
            // key 加上 l
            $arr[$key+$l] = $val;
        }

        echo sprintf("l=%s, mid=%s, r=%s\n", $l, $mid, $r);
    }
}

$m = new MergeSort();
$arr = [5, 4, 3, 2, 1, 7];
$m->sort($arr);

print_r($arr);

File: sort/MergeSort.go
package main
import "fmt"

func sort(arr []int) {
    interally(arr, 0, len(arr)-1)
}

// 递归分解
func interally(arr []int, l int, r int) {
    // 递归终止条件
    if l >= r {
        return 
    }
    var mid int = l + (r-l) / 2 // 这样写最大地好处就是可以处理大数溢出地概率会减小
    interally(arr, l, mid)
    interally(arr, mid+1, r)
    merge(arr, l, mid, r)
}

// 合并治理
func merge(arr []int, l int, m int, r int) {
    var i = l; // 左侧起始位置
    var j = m + 1; // 右侧起始位置
    var tmp []int
    for i <= m && j <= r {
        if arr[i] < arr [j] {
            tmp = append(tmp, arr[i])
            i++
        } else {
            tmp = append(tmp, arr[j])
            j++
        }
    }

    // 如果右侧还没有处理完
    if j <= r {
        // 右侧没有处理完成，说明左侧一定处理完了
        i = j 
        m = r
    }
    for ; i <= m; i++ {
        tmp = append(tmp, arr[i])
    }
    // 将数据复制到 arr 中
    for i := 0; i < len(tmp); i++ {
        arr[l+i] = tmp[i]
    }
}

func main()  {
	
	var arr []int = []int{6, 7, 1, 3, 2, 5, 4}

    sort(arr)
    
    fmt.Println(arr)

}